renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -425,6 +444,680 @@ ShaderDebugTrace D3D12Replay::DebugPixel(uint32_t eventId, uint32_t x, uint32_t
+  descRange.RangeType = D3D12_DESCRIPTOR_RANGE_TYPE_UAV;
+  descRange.NumDescriptors = 1;
+  descRange.BaseShaderRegister = 0;
+  descRange.RegisterSpace = 0;
+  descRange.Flags = D3D12_DESCRIPTOR_RANGE_FLAG_NONE;
+  descRange.OffsetInDescriptorsFromTableStart = 0;
+
+  modsig.Parameters.push_back(D3D12RootSignatureParameter());
+  D3D12RootSignatureParameter &param = modsig.Parameters.back();
+  param.ParameterType = D3D12_ROOT_PARAMETER_TYPE_DESCRIPTOR_TABLE;
+  param.ShaderVisibility = D3D12_SHADER_VISIBILITY_PIXEL;
+  param.DescriptorTable.NumDescriptorRanges = 1;
+  param.DescriptorTable.pDescriptorRanges = &descRange;
+
+  uint32_t sigElem = uint32_t(modsig.Parameters.size() - 1);
+
+  // Create the root signature for gathering initial pixel shader values
+  ID3DBlob *root = m_pDevice->GetShaderCache()->MakeRootSig(modsig);
+  ID3D12RootSignature *pRootSignature = NULL;
+  hr = m_pDevice->CreateRootSignature(0, root->GetBufferPointer(), root->GetBufferSize(),
+                                      __uuidof(ID3D12RootSignature), (void **)&pRootSignature);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create root signature for pixel shader debugging HRESULT: %s",
+           ToStr(hr).c_str());
+    SAFE_RELEASE(root);
+    SAFE_RELEASE(psBlob);
+    SAFE_RELEASE(pInitialValuesBuffer);
+    return empty;
+  }
+  SAFE_RELEASE(root);
+
+  WrappedID3D12PipelineState *origPSO =
+      m_pDevice->GetResourceManager()->GetCurrentAs<WrappedID3D12PipelineState>(rs.pipe);
+
+  RDCASSERT(origPSO->IsGraphics());
+
+  D3D12_EXPANDED_PIPELINE_STATE_STREAM_DESC pipeDesc;
+  origPSO->Fill(pipeDesc);
+
+  // All PSO state is the same as the event's, except for the pixel shader and root signature
+  pipeDesc.PS.BytecodeLength = psBlob->GetBufferSize();
+  pipeDesc.PS.pShaderBytecode = psBlob->GetBufferPointer();
+  pipeDesc.pRootSignature = pRootSignature;
+
+  ID3D12PipelineState *initialPso = NULL;
+  hr = m_pDevice->CreatePipeState(pipeDesc, &initialPso);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create PSO for pixel shader debugging HRESULT: %s", ToStr(hr).c_str());
+    SAFE_RELEASE(psBlob);
+    SAFE_RELEASE(pInitialValuesBuffer);
+    SAFE_RELEASE(pRootSignature);
+    return empty;
+  }
+
+  // Add the descriptor for our UAV
+  size_t prevHeapCount = rs.heaps.size();
+  std::set<ResourceId> copiedHeaps;
+  PortableHandle shaderDebugUav = ToPortableHandle(GetDebugManager()->GetCPUHandle(SHADER_DEBUG_UAV));
+  AddDebugDescriptorToRenderState(m_pDevice, rs, shaderDebugUav,
+                                  D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV, sigElem, copiedHeaps);
+
+  // Set the PSO and root signature
+  rs.pipe = GetResID(initialPso);
+  rs.graphics.rootsig = GetResID(pRootSignature);
+
+  ID3D12GraphicsCommandListX *cmdList = m_pDevice->GetDebugManager()->ResetDebugList();
+
+  // If we needed to add a heap for our debug UAV, update the state on the command list
+  if(rs.heaps.size() != prevHeapCount)
+    rs.ApplyDescriptorHeaps(cmdList);
```

Comment:
This condition is wrong. `ClearUnorderedAccessViewUint` requires that the descriptor heap containing the UAV is set as active with `SetDescriptorHeaps`. Whether we've changed the number of heaps is irrelevant, we must set the heap that we created the UAV on in all cases.

I'm not meaning to patronise here but since this seemed to cause confusion in another case I'll explain: In D3D12 the state on a command list (whether that's the bound descriptor heaps/bound pipeline/anything else) is _local_ to that command list. Meaning that the state one one command list has no interaction at all with the state on another command list, there's no inheritance or persistent state between command lists. Once you close a command list and submit it that state is effectively gone, the next one you set up needs to re-set the state from scratch. (I'm ignoring bundles here since no-one seems to use them anyway).

This is different to D3D11 where the state remains present on the immediate context until it's changed, so in RenderDoc's case if you do some rendering and change state you might need to restore it again.

The `m_RenderState` which you're modifying is the current saved state wherever we are mid-way through a command list. We track this internally and it's used so that we can fake state inheritance into any command lists we want, e.g. if we want to replay the draw (like you do below with the call to `ReplayLog`) we can restore the state.

In this case the command list is only doing one thing - clearing the UAV. So all we need to do is set the state needed for that, which is setting the descriptor heap. Similarly this is why the previous command list you had that applied the state then closed and executed did nothing, because setting state in a command list doesn't have any side-effects outside of that command list. Restoring the contents of `m_RenderState` to `prevstate` _is_ important because that could be used subsequently for examining the current state, up to the point where we do another full replay to a different event and reset our tracking.

Reply:
Thanks for the verbose explanation. Wish it was spelled out as clearly as that on MSDN. :)